### PR TITLE
Fixes #15134: Migrate ncf write technique api to Rudder

### PR DESCRIPTION
**Pull request by VinceMacBuche**
https://issues.rudder.io/issues/15134

### CODE REVIEW

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala [hunk 1]
```diff
@@ -173,18 +180,18 @@ class TechniqueWriter (
     for {
       agentFiles <- writeAgentFiles(technique, methods, modId, committer)
       metadata   <- writeMetadata(technique, methods, modId, committer)
+      commit     <- archiver.commitTechnique(technique,metadata +: agentFiles, modId, committer, s"Committing technique ${technique.name}")
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala]: I feel like their should be some kind of guards (semaphor) around our git changes, perhaps a lock by directory or technique... 
At least the commit is selective in the files it commits. We can keep that for now.

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala [hunk 2]
```diff
@@ -197,55 +204,196 @@ class TechniqueWriter (
...
+    for {
+      tech <- t
+      repo <- t2
+    } yield {
+      tech +: repo
+    }
+  }
   def agentMetadata ( technique : Technique, methods : Map[BundleName, GenericMethod] )  : PureResult[NodeSeq] = {
-    // We need to add reporting bundle if there is a method call that does not support cfengine (agent support does not contains both cfe agent)
+    // We need to add reporting bundle if there is a method call that does not supporval t cfengine (agent support does not contains both cfe agent)
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala]: typo

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala [hunk 3]
```diff
@@ -404,26 +553,40 @@ class DSCTechniqueWriter(
...
 class TechniqueArchiverImpl (
     override val gitRepo                   : GitRepositoryProvider
-  , override val gitRootDirectory          : File
+  , override val gitRootDirectory          : JFile
   , override val xmlPrettyPrinter          : RudderPrettyPrinter
   , override val relativePath              : String
   , override val gitModificationRepository : GitModificationRepository
   , personIdentservice : PersonIdentService
-) extends GitArchiverUtils with TechniqueArchiver{
+) extends GitArchiverUtils with TechniqueArchiver with Loggable {
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala]: why adding loggable?

File: webapp/sources/rudder/rudder-core/src/test/resources/configuration-repository/techniques/ncf_techniques/technique_by_Rudder/1.0/metadata.xml [hunk 4]
```diff
@@ -31,7 +31,7 @@
     <SECTION component="true" multivalued="true" name="Command execution">
       <REPORTKEYS>
         <VALUE>Write-Host &quot;testing special characters ` è &amp;é 'à é &quot;</VALUE>
-        <VALUE>/bin/echo &quot;testing special characters ` è &amp;é 'à é &quot;</VALUE>
+        <VALUE>/bin/echo &quot;testing special characters ` è &amp;é 'à é &quot;\</VALUE>
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/test/resources/configuration-repository/techniques/ncf_techniques/technique_by_Rudder/1.0/metadata.xml]: why '\' ?

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/RestExtractorService.scala [hunk 5]
```diff
@@ -855,10 +866,10 @@ case class RestExtractorService (
       expirationValue  <- extractJsonString(json, "expirationDate", DateFormaterService.parseDate)
       authType    <- extractJsonString(json, "authorizationType", ApiAuthorizationKind.parse)
 
-      acl     <- extractJsonArray(json \ "acl")(
+      acl     <- extractJsonArray(json , "acl")(
                        // I need an option here, but this function does not need to return an option
                        // AndThern it to wrap the result in a option, then unwrap it with a getOrElse after it ran
-                       (extractApiACLFromJSON _ ).andThen(_.map(Some(_)))
+                       (extractApiACLFromJSON _ )/*.andThen(_.map(Some(_)))*/
```
Comment (fanf) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/RestExtractorService.scala]: that code and the comment above seems to be no more used

**Comment by VinceMacBuche** [comment]
Commit modified

File: webapp/sources/rudder/rudder-core/src/test/resources/configuration-repository/techniques/ncf_techniques/technique_by_Rudder/1.0/metadata.xml (hunk 4 above)
Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/test/resources/configuration-repository/techniques/ncf_techniques/technique_by_Rudder/1.0/metadata.xml]: Adding support of \ at the end of a value ! not breaking generated promises anymore

**Comment by VinceMacBuche** [comment]
Commit modified

**Comment by VinceMacBuche** [comment]
Commit modified

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala [hunk 6]
```diff
@@ -218,33 +214,168 @@ class TechniqueWriter (
...
+  // We need to add a reporting bundle for this method to generate a na report for any method with a condition != any/cfengine (which ~= true
+  def methodNeedReporting(method : MethodCall) =  method.condition != "any" && method.condition != "cfengine-community"
+  def needReportingBundle(technique : Technique) = technique.methodCalls.exists(methodNeedReporting)
+
+  def canonifyCondition(methodCall: MethodCall) = {
+    methodCall.condition.replaceAll("""(\$\{[^\}]*})""","""",canonify("$1"),"""")
+  }
+
+  // regex to match double quote characters not preceded by a backslash, and backslash not preceded by backslash or not followed by a backslash or a quote (simple or double)
+  def escapeCFEngineString(value : String ) = value.replaceAll("""(?<!\\)"""", """\\"""").replaceAll("""(?<!\\)\\(?!(\\|"|'))""","""\\\\""")
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala]: This seems very brittle and a good candidate for a parser. We should create an issue for that.

File: webapp/sources/rudder/rudder-web/src/main/scala/bootstrap/liftweb/checks/CheckNcfTechniqueUpdate.scala [hunk 7]
```diff
@@ -120,13 +116,14 @@ class CheckNcfTechniqueUpdate(
...
                               }
         // First get generic methods
-        methodsResponse    <- tryo ( methodsRequest.asString , "An error occurred while fetching generic methods from ncf API", NcfApiRequestFailed)
-        methodsResult      <- if(methodsResponse.isSuccess) {
-                                Right(methodsResponse.body)
-                              } else {
-                                Left(NcfApiRequestFailed(s"Failure getting generic methods from ncf api: code ${methodsResponse.code}: ${methodsResponse.body}", None))
-                              }
-        methods            <- (parse(methodsResult) \ "data" \ "generic_methods" match {
+        // Then take care of techniques
```
Comment (fanf) [inline webapp/sources/rudder/rudder-web/src/main/scala/bootstrap/liftweb/checks/CheckNcfTechniqueUpdate.scala]: the comment is incorrect here

**Comment by VinceMacBuche** [comment]
Commit modified

**Comment by VinceMacBuche** [comment]
Commit modified

**Pull request merged**